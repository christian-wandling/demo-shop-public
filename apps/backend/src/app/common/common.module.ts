import { Module } from '@nestjs/common';
import { PrismaService } from './services/prisma.service';
import { MonitoringService } from './services/monitoring.service';

@Module({
  providers: [PrismaService, MonitoringService],
  exports: [PrismaService, MonitoringService],
})
export class CommonModule {}
