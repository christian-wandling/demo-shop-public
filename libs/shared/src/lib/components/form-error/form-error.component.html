@if (errors()?.['required']) {
  <div>{{ fieldName() }} is required</div>
}

@if (errors()?.['email']) {
  <div>Email invalid</div>
}
