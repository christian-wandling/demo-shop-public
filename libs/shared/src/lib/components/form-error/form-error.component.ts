import { ChangeDetectionStrategy, Component, input } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ValidationErrors } from '@angular/forms';

@Component({
  selector: 'lib-form-error',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './form-error.component.html',
  styleUrl: './form-error.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: {
    'class': 'text-xs text-red-600',
  },
})
export class FormErrorComponent {
  readonly fieldName = input<string>('Field');
  readonly errors = input.required<ValidationErrors | null>();
}
