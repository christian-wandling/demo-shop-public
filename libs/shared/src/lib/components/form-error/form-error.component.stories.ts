import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { FormErrorComponent } from './form-error.component';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { CommonModule } from '@angular/common';
import { importProvidersFrom } from '@angular/core';

const meta: Meta<FormErrorComponent> = {
  component: FormErrorComponent,
  title: 'Shared/FormErrorComponent',
  decorators: [
    applicationConfig({
      providers: [importProvidersFrom(CommonModule)],
    }),
  ],
  argTypes: {
    fieldName: {
      control: {
        type: 'text',
      },
    },
  },
};
export default meta;
type Story = StoryObj<FormErrorComponent>;

export const RequiredError: Story = {
  args: {
    errors: { required: true },
  },
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText(`${args.fieldName} is required`)).toBeTruthy();
  },
};

export const InvalidEmailError: Story = {
  args: {
    errors: { email: true },
  },
  parameters: {
    controls: {
      exclude: ['fieldName'],
    },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText('Email invalid')).toBeTruthy();
  },
};

export const MultipleErrors: Story = {
  args: {
    fieldName: 'Name',
    errors: { required: true, email: true },
  },
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText(`${args.fieldName} is required`)).toBeTruthy();
    expect(canvas.getByText('Email invalid')).toBeTruthy();
  },
};
