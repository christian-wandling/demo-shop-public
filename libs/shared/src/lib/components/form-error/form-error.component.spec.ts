import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormErrorComponent } from './form-error.component';
import { CommonModule } from '@angular/common';

describe('FormErrorComponent', () => {
  let component: FormErrorComponent;
  let fixture: ComponentFixture<FormErrorComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [CommonModule, FormErrorComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(FormErrorComponent);
    component = fixture.componentInstance;
    fixture.componentRef.setInput('errors', null);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should have default field name as "Field"', () => {
    expect(component.fieldName()).toBe('Field');
  });

  it('should display custom field name', () => {
    fixture.componentRef.setInput('fieldName', 'Email');
    fixture.detectChanges();
    expect(component.fieldName()).toBe('Email');
  });

  it('should show required error message when required error is present', () => {
    fixture.componentRef.setInput('fieldName', 'Username');
    fixture.componentRef.setInput('errors', { required: true });

    fixture.detectChanges();
    const compiled = fixture.nativeElement;
    expect(compiled.textContent).toContain('Username is required');
  });

  it('should show email error message when email error is present', () => {
    fixture.componentRef.setInput('errors', { email: true });

    fixture.detectChanges();
    const compiled = fixture.nativeElement;
    expect(compiled.textContent).toContain('Email invalid');
  });

  it('should show all applicable error messages when multiple errors are present', () => {
    fixture.componentRef.setInput('fieldName', 'Email');
    fixture.componentRef.setInput('errors', {
      required: true,
      email: true,
    });

    fixture.detectChanges();
    const compiled = fixture.nativeElement;
    expect(compiled.textContent).toContain('Email is required');
    expect(compiled.textContent).toContain('Email invalid');
  });

  it('should not show any error messages when errors is null', () => {
    fixture.componentRef.setInput('errors', null);

    fixture.detectChanges();
    const compiled = fixture.nativeElement;
    expect(compiled.textContent).toBe('');
  });

  it('should have the correct host classes', () => {
    const compiled = fixture.nativeElement;
    expect(compiled.classList.contains('text-xs')).toBeTruthy();
    expect(compiled.classList.contains('text-red-600')).toBeTruthy();
  });

  describe('Snapshots', () => {
    it('should match snapshot with required error', () => {
      fixture.componentRef.setInput('fieldName', 'Username');
      fixture.componentRef.setInput('errors', { required: true });

      fixture.detectChanges();
      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should match snapshot with email error', () => {
      fixture.componentRef.setInput('fieldName', 'Email');
      fixture.componentRef.setInput('errors', { email: true });

      fixture.detectChanges();
      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should match snapshot with multiple errors', () => {
      fixture.componentRef.setInput('fieldName', 'Email');
      fixture.componentRef.setInput('errors', {
        required: true,
        email: true,
      });

      fixture.detectChanges();
      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should match snapshot with no errors', () => {
      fixture.componentRef.setInput('fieldName', 'Email');
      fixture.componentRef.setInput('errors', null);

      fixture.detectChanges();
      expect(fixture.nativeElement).toMatchSnapshot();
    });
  });
});
