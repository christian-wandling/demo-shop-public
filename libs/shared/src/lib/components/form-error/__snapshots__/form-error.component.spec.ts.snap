// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormErrorComponent Snapshots should match snapshot with email error 1`] = `
<div
  class="text-xs text-red-600"
  id="root9"
>
  
  <div>
    Email invalid
  </div>
  
</div>
`;

exports[`FormErrorComponent Snapshots should match snapshot with multiple errors 1`] = `
<div
  class="text-xs text-red-600"
  id="root10"
>
  <div>
    Email is required
  </div>
  
  <div>
    Email invalid
  </div>
  
</div>
`;

exports[`FormErrorComponent Snapshots should match snapshot with no errors 1`] = `
<div
  class="text-xs text-red-600"
  id="root11"
>
  
  
</div>
`;

exports[`FormErrorComponent Snapshots should match snapshot with required error 1`] = `
<div
  class="text-xs text-red-600"
  id="root8"
>
  <div>
    Username is required
  </div>
  
  
</div>
`;
